script::Module as a view.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21853 Add interface declarations to JIT
* **#21814 script::Module as a view.**

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

Differential Revision: [D15839583](https://our.internmc.facebook.com/intern/diff/D15839583)

zdevito committed:
script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -410,96 +394,65 @@ struct TORCH_API Module {
   }
-  at::optional<size_t> find_offset(
+
+  void get_or_add_slot(
```
This seems more like `set_or_add_slot`

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -142,70 +129,44 @@ struct TORCH_API Module {
+        name, module->type(), module->module_object(), EntityType::MODULE);
   }
 
   void set_parameter(const std::string& name, at::Tensor v) {
```
(future PR) we have a `set_parameter` but no `set_buffer`/`set_attribute`. Probably worth adding

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -142,70 +129,44 @@ struct TORCH_API Module {
     return get_method("forward")(std::move(inputs));
   }
-
   void register_buffer(const std::string& name, autograd::Variable v) {
```
May be worth dropping a note here about why buffers are treated as attributes

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -222,53 +183,71 @@ struct TORCH_API Module {
-  }
-  c10::ArrayRef<Slot> get_attributes() const {
-    return attributes_;
+  std::vector<std::shared_ptr<Module>> get_modules() const {
```
I don't think is makes sense for these to be `shared_ptr` anymore, since Module is a value type anyway

zdevito replied:
I agree, but this is a big breaking API change for anyone using this API. Any call like `foo.get_submodule("bar")->is_training()` will need to change because the thing being returned is no longer a pointer. We should do it but it will need to be in a follow up PR to minimize the amount of API changes this one causes.

suo commented on torch/csrc/jit/script/module.h:
```diff
@@ -222,53 +183,71 @@ struct TORCH_API Module {
-    return nullptr;
+    return c10::nullopt;
   }
   std::shared_ptr<Module> find_module(const std::string& name) const {
```
Same here

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -449,17 +450,17 @@ void initJitScriptBindings(PyObject* module) {
           "_has_attribute",
           [](Module& self, const std::string& name) -> bool {
-            return self.find_attribute(name);
+            return !!self.find_attribute(name);
```
has_value?

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

We've changed the implementation of iterating over parameters/attributes/modules
many times, and each time we've had to slightly break backward compatibility.
This also changes the api to use a class (slot_list/iterator) to encapsulate
iteration of parameters/attributes/modules. This should allow us to make
implementation changes without affecting the user API.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

We've changed the implementation of iterating over parameters/attributes/modules
many times, and each time we've had to slightly break backward compatibility.
This also changes the api to use a class (slot_list/iterator) to encapsulate
iteration of parameters/attributes/modules. This should allow us to make
implementation changes without affecting the user API.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

We've changed the implementation of iterating over parameters/attributes/modules
many times, and each time we've had to slightly break backward compatibility.
This also changes the api to use a class (slot_list/iterator) to encapsulate
iteration of parameters/attributes/modules. This should allow us to make
implementation changes without affecting the user API.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

We've changed the implementation of iterating over parameters/attributes/modules
many times, and each time we've had to slightly break backward compatibility.
This also changes the api to use a class (slot_list/iterator) to encapsulate
iteration of parameters/attributes/modules. This should allow us to make
implementation changes without affecting the user API.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

We've changed the implementation of iterating over parameters/attributes/modules
many times, and each time we've had to slightly break backward compatibility.
This also changes the api to use a class (slot_list/iterator) to encapsulate
iteration of parameters/attributes/modules. This should allow us to make
implementation changes without affecting the user API.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head

zdevito committed:
Update on "script::Module as a view."

script::Module as a view.

This moves all the module description logic (i.e. what members
of the class are parameters) into the ClassType for the module
itself. This allows us to remove all state in script::Module,
making it a stateless view of the class.

We've changed the implementation of iterating over parameters/attributes/modules
many times, and each time we've had to slightly break backward compatibility.
This also changes the api to use a class (slot_list/iterator) to encapsulate
iteration of parameters/attributes/modules. This should allow us to make
implementation changes without affecting the user API.

gh-metadata: pytorch pytorch 21814 gh/zdevito/65/head
